Sources/Target/Linux/Process.cpp

```diff
@@ -212,20 +213,58 @@ ErrorCode Process::wait() {
     _currentThread->updateStopInfo(status);
+    didSpawnThread =
+        _currentThread->_stopInfo.reason == StopInfo::kReasonThreadSpawn;
 
     switch (_currentThread->_stopInfo.event) {
     case StopInfo::kEventNone:
-      // If the thread is stopped for no reason, it means the debugger (ds2)
-      // sent a SIGSTOP to it while it was already stopped for another reason,
-      // the SIGSTOP was queued, and now we're getting this notification. We
-      // have to ignore this and just let the thread continue.
-      // If `stepping` is true, it means the thread was actually being
-      // single-stepped before stopping, so instead of doing a `resume()`, we
-      // have to do a new `step()`.
-      if (stepping) {
+      // If _stopInfo.event is kEventNone it means the thread is either
+      // 1. stopped because another thread of the inferior hit a breakpoint
+      //    and ds2 sent a SIGSTOP to pause this thread while the breakpoint
+      //    is evaluated. In this case we just resume the thread.
+      // 2. stopped because this thread called clone(2) (or similar e.g.
+      //    Pthread_spawn(2)). In this case we have two subcases:
+      //    a. The thread was running and hit the clone/spawn and thus was
+      //       send a SIGTRAP and stopped. We just resume it and goto
+      //       continue_waiting. The loop will continue and the next waitpid
+      //       will handle the freshly spawned thread. In this case we call
+      //       resume as well.
+      //    b. We were single stepping (c.f. the `stepping` var) and
+      //       were stopped by a SIGTRAP for the step while simultaneously
+      //       being stopped via the SIGTRAP for the clone/spawn.
+      //       In this case we can run into a race condition with the
+      //       waitpid(3) call and thus have to handle the cloned thread
+      //       and the currentThread at once. So we call step on the
+      //       _currentThreadand then beforeResume the cloned thread.
+      //
+      // We also have another theoretically possible case where we are not
+      // stopped by a clone but still stepping while encountering a kEventNone.
+      // This should not happen and we just assert that it does not.
+      //
+
+      if (didSpawnThread && stepping) { //(2b)
+        unsigned long spawnedThreadTID = 0;
+        auto error = ptrace().getSpawnedTID(_pid, spawnedThreadTID);
+        if (error != 0) {
+          return error;
+        }
+
+        int spawnedThreadStatus = 0;
+        spawnedThreadTID =
+            ::waitpid(spawnedThreadTID, &spawnedThreadStatus, WNOHANG | __WALL);
+        DS2LOG(Debug, "child thread tid %lu %s", spawnedThreadTID,
+               Stringify::WaitStatus(spawnedThreadStatus));
         _currentThread->step();
+        auto newThread = new Thread(this, spawnedThreadTID);
+
+        newThread->beforeResume();
+      } else if (stepping) {
+        DS2BUG("We should never see a case where we're\n"
```

Comment:
Multi-line logs will fuck up the output format. Make it something more concise like
`DS2BUG("Inconsistent thread stop info");`

Reply:
Fixed.